jvm-packages/xgboost4j/src/native/xgboost4j.cpp

```diff
@@ -661,6 +662,84 @@ JNIEXPORT jint JNICALL Java_ml_dmlc_xgboost4j_java_XGBoostJNI_XGBoosterPredict
   return ret;
 }
 
+/*
+ * Class:     ml_dmlc_xgboost4j_java_XGBoostJNI
+ * Method:    XGBoosterPredictFromDense
+ * Signature: (J[FJJFIII[F[[F)I
+ */
+JNIEXPORT jint JNICALL Java_ml_dmlc_xgboost4j_java_XGBoostJNI_XGBoosterPredictFromDense(
+    JNIEnv *jenv, jclass jcls, jlong jhandle, jfloatArray jdata, jlong num_rows, jlong num_features,
+    jfloat missing, jint iteration_begin, jint iteration_end, jint predict_type,
+    jfloatArray jmargin, jobjectArray jout) {
+  API_BEGIN();
+  BoosterHandle handle = reinterpret_cast<BoosterHandle>(jhandle);
+
+  /**
+   * Create array interface.
+   */
+  namespace linalg = xgboost::linalg;
+  jfloat *data = jenv->GetFloatArrayElements(jdata, nullptr);
+  auto t_data = linalg::MakeTensorView(
+      xgboost::Context::kCpuId,
+      xgboost::common::Span{data, static_cast<std::size_t>(num_rows * num_features)}, num_rows,
+      num_features);
+  auto s_array = linalg::ArrayInterfaceStr(t_data);
+
+  /**
+   * Create configuration object.
+   */
+  xgboost::Json config{xgboost::Object{}};
+  config["cache_id"] = xgboost::Integer{};
+  config["type"] = xgboost::Integer{static_cast<std::int32_t>(predict_type)};
+  config["iteration_begin"] = xgboost::Integer{static_cast<xgboost::bst_layer_t>(iteration_begin)};
+  config["iteration_end"] = xgboost::Integer{static_cast<xgboost::bst_layer_t>(iteration_end)};
+  config["missing"] = xgboost::Number{static_cast<float>(missing)};
+  config["strict_shape"] = xgboost::Boolean{true};
+  std::string s_config;
+  xgboost::Json::Dump(config, &s_config);
+
+  /**
+   * Handle base margin
+   */
+  BoosterHandle proxy{nullptr};
```

Comment:
Please help add a test for prediction with margin. With regression using `reg:squarederror`:
```
pred0 = booster.inplace_predict(X, margin)
pred1 = booster.inplace_predict(X) + margin
assert pred0 == pred1
```

Reply:
@trivialfis Hello! I wrote a test, but it appears to be failing. I think it may be related to this: https://github.com/dmlc/xgboost/pull/9536

Is the in-place prediction not working properly with margins at the moment?

Reply:
The bug happens when you are running prediction with base margin on GPU while the input is from the CPU or the other way around. I don't think it's relevant to this PR.

Reply:
Feel free to push your tests, I can take a look.

Reply:
Thank you very much! I pushed the new tests. Let me know if you have any questions on anything! Right now what I'm seeing with the failing test is that the prediction( x with margin) != prediction(x) + margin

Reply:
Forgot to tell you the `base_score` needs to be set. Apologies.

Reply:
Awesome! That did the trick. I'll post your changes from below shortly.

Reply:
Great! Please merge/rebase the latest branch.

Reply:
Done! Let me know if there's anything else needed!

Reply:
> Done! Let me know if there's anything else needed!

Fix CI errors, if there's any. ;-)

Reply:
> 

It looks like the CI failed with some of the new C++ code you added from the sovrn#9 branch. I tried taking a look at it, but by no means am I a C++ expert. It looks like some sort of parameter type mismatch is occurring. Could you take a look?

Reply:
@trivialfis 

I think maybe this could fix the problem:

`/**
 * Create array interface.
 */
namespace linalg = xgboost::linalg;
jfloat *data = jenv->GetFloatArrayElements(jdata, nullptr);
xgboost::HostDeviceVector<float> hdv(data, data + num_rows * num_features, xgboost::Context::kCpuId);
auto t_data = linalg::MakeTensorView(xgboost::Context::kCpuId, &hdv, num_rows, num_features);
auto s_array = linalg::ArrayInterfaceStr(t_data);`

But I'm not certain on the memory management here. Any thoughts?